runners/spark/src/main/java/org/apache/beam/runners/spark/util/GlobalWatermarkHolder.java

```diff
@@ -0,0 +1,200 @@
+
+import com.google.common.annotations.VisibleForTesting;
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.broadcast.Broadcast;
+import org.apache.spark.streaming.api.java.JavaStreamingContext;
+import org.apache.spark.streaming.api.java.JavaStreamingListener;
+import org.apache.spark.streaming.api.java.JavaStreamingListenerBatchCompleted;
+import org.joda.time.Instant;
+
+
+/**
+ * A {@link Broadcast} variable to hold the global watermarks for a micro-batch.
+ *
+ * <p>For each source, holds a queue for the watermarks of each micro-batch that was read,
+ * and advances the watermarks according to the queue (first-in-first-out).
+ */
+public class GlobalWatermarkHolder {
+  // the broadcast is broadcasted to the workers.
+  private static volatile Broadcast<Map<Integer, SparkWatermarks>> broadcast = null;
+  // this should only live in the driver so transient.
+  private static final transient Map<Integer, Queue<SparkWatermarks>> sourceTimes = new HashMap<>();
+
+  public static void add(int sourceId, SparkWatermarks sparkWatermarks) {
+    Queue<SparkWatermarks> timesQueue = sourceTimes.get(sourceId);
+    if (timesQueue == null) {
+      timesQueue = new ConcurrentLinkedQueue<>();
+    }
+    timesQueue.offer(sparkWatermarks);
+    sourceTimes.put(sourceId, timesQueue);
+  }
+
+  @VisibleForTesting
+  public static void addAll(Map<Integer, Queue<SparkWatermarks>> sourceTimes) {
+    for (Map.Entry<Integer, Queue<SparkWatermarks>> en: sourceTimes.entrySet()) {
+      int sourceId = en.getKey();
+      Queue<SparkWatermarks> timesQueue = en.getValue();
+      while (!timesQueue.isEmpty()) {
+        add(sourceId, timesQueue.poll());
+      }
+    }
+  }
+
+  /**
+   * Returns the {@link Broadcast} containing the {@link SparkWatermarks} mapped
+   * to their sources.
+   */
+  public static Broadcast<Map<Integer, SparkWatermarks>> get() {
+    return broadcast;
+  }
+
+  /**
+   * Advances the watermarks to the next-in-line watermarks.
+   * SparkWatermarks are monotonically increasing.
+   */
+  public static void advance(JavaSparkContext jsc) {
+    synchronized (GlobalWatermarkHolder.class){
+      if (sourceTimes.isEmpty()) {
+        return;
+      }
+
+      // update all sources' watermarks into the new broadcast.
+      Map<Integer, SparkWatermarks> newBroadcast = new HashMap<>();
+
+      for (Map.Entry<Integer, Queue<SparkWatermarks>> en: sourceTimes.entrySet()) {
+        if (en.getValue().isEmpty()) {
+          continue;
+        }
+        Integer sourceId = en.getKey();
+        Queue<SparkWatermarks> timesQueue = en.getValue();
+
+        // current state, if exists.
+        Instant currentLowWatermark = BoundedWindow.TIMESTAMP_MIN_VALUE;
```

Comment:
I've updated the design doc: https://docs.google.com/document/d/1bOu0-g7Z1LVG0kYEQ38GDWaFjFl45u-2aZV87RmlLKI/edit?usp=sharing